### ISSUE TITLE
Light2D Mix/Mask mode issues whith other lights

### ISSUE THREAD

**Issue by mralve** [godotengine/godot#8685]
Windows 10 x64 - Gtx 680 - GD v.3.0 x64

**Hello, Godot devs...**
I have come across a very make or break issue for me in Godot, and that is when I use a light2D with the blending mode Mix or Mask and if that light2D gets close to some or sometimes all of them, light2D's then it will randomly turn off / or get cut in half... Now if I re-open the scene all the lights who is in or close to the light2D with mix/mask mode will be randomly on or of. I've been experiencing this since 2.1.x and up to now on my main win 10 computer and on my win 10 x86 laptop in the editor or exported version of the project.

Sorry if this is a bad explanation :P.
Now if this is something I'm doing wrong or if it is a noob issue as I'm new to Godot please correct me, but I strongly doubt that it is my wrong doing. 

**How I reproduce this:**
- Create at least one light2D with mask/mix mode
- Create some normal light2Ds with add mode
- Move around the light with mix mode over the other lights, if all is working as it shud try to re-open the level several times and that shud make some lights behave badly.

**Example project:**
https://www.dropbox.com/s/f0und6dkzzugxx2/Light2DModes_Project.zip?dl=0

**Comment by eon-s** [comment]
Related or maybe the same as #7485 on 2.x

I think Light2D was not touched since the migration to the new renderer thus it has the same and more bugs as before, I hope CanvasItems get some love before the 3 alpha proper.

**Comment by mralve** [comment]
Yes, absolutely it's a very big turn off for me in Godot. I love the engine and community so more love to canvasItems 👍

**Comment by reduz** [comment]
This is because lighting in godot works with multiple passes in 2D, should be changed to single pass but it's a good amount of work, so kicking to 3.1

**Comment by Zylann** [comment]
Getting issues with this too in 3.0 alpha2, I just can't have more than one light set as mask. If I do, they just all turn off, or I can see only their intersection somehow.

**Comment by reduz** [comment]
This will get fixed when single pass 2D lighting is implemented, ran out of
time to do this for 3.0

**Comment by reduz** [comment]
This is impossible to fix in 3.1, will have to wait for Vulkan and single pass rendering, so kicking to next version.

**Comment by akien-mga** [comment]
See also #7485 for related Light2D issues (potentially duplicate, didn't investigate fully). Both bug reports should be handled together eventually.

**Comment by giulianob** [comment]
Chiming in on the issue. I read that there was concern with doing lighting in a single pass on GLES2 because it was have to be rendered to an internal atlas which means a limit on the number of lights and/or size. I personally think that's a totally acceptable trade-off (maybe it's configurable) because right now 2d lighting isnt very useful on mobile. Im planning on implementing my own atlas essentially to workaround it for mobile and would love to use the built in lighting if possible.

**Comment by eon-s** [comment]
@giulianob at the current stable version you may be able to use a dynamic texture on lights to adjust the mask, a viewport texture or something like that.

**Comment by giulianob** [comment]
How does that help it so the lights are rendered in a single pass?

**Comment by eon-s** [comment]
Your mask will be on a single light, the problem is with many lights in mask mode.

**Comment by giulianob** [comment]
I was mainly talking about performance of light because another issue linked to this one

**Comment by akien-mga** [comment]
Still reproducible in the current master branch (966c68ba) with project from #22981:
[light2d_as_mask.zip](https://github.com/godotengine/godot/files/2475384/light2d_as_mask.zip)

According to @reduz's comment above, this should be fixed in 4.0 but will likely stay broken on 3.2 (unless someone else finds a way to solve it with refactoring the whole 3.x rendering).

**Comment by eon-s** [comment]
It may remain broken for GLES2 on 4.x too.

**Comment by akien-mga** [comment]
Another reproduction project from #14361:
https://github.com/godotengine/godot/files/1537745/normal_mapped_lighting_bug.zip

See also GIFs there.

**Comment by mralve** [comment]
Wow did not expect this issue to hang in as long as it did, interesting. I don't even remember what that project was. : )

**Comment by guelinator** [comment]
We are having the same issue here. I am thinking about baning Light2D, because it really can drive one craz.
A designer uses a light2d in a scene.
The scene gets instanciated multiple times.
Everything works fine, until the Light2D's touch.

Makes everything very unpredictable, and really cumbersome.

Even if it gets fixed in vulkan, HTML export will suffer.

**Comment by Vadimledyaev** [comment]
Will this be fixed in v 4.0 / Vulkan?

It is vital for me because I'm currently developing a commercial project with a must-have 2D light in mask mode. 

Thanks!

**Comment by Calinou** [comment]
Related to #33657.

**Comment by vyshliy** [comment]
Confirm similar behavior. The problem arises when one Light2D (mask mode) overlaps with another with the same Cull Mask value. I offer an example project for demonstration. 

https://user-images.githubusercontent.com/72736632/180650722-b36a25b0-be5b-48ec-b774-b10f6c10d8ac.mp4


[Light2Ds_overlap_issue.zip](https://github.com/godotengine/godot/files/9175948/Light2Ds_overlap_issue.zip)

**Comment by vyshliy** [comment]
Can confirm similar behavior with 3.4.4. The issue arises when one Light2D (mask mode) overlaps with another with the same Cull Mask value. Play attached example project for demonstration.
[Light2Ds_overlap_issue.zip](https://github.com/godotengine/godot/files/9175972/Light2Ds_overlap_issue.zip)

https://user-images.githubusercontent.com/72736632/180651528-8721f294-dfb4-4104-a9d8-6ee9e7dd5981.mp4

**Comment by clayjohn** [comment]
Bumping to 3.x as 4.0 has removed the light mask mode and replaced it with the clip_children option and the CanvasGroup

**Comment by hsandt** [comment]
While creating an MRP for another issue I stumbled on this. So I thought I might as well share my MRP if you want to test more specifically what happens where you have a rectangle that reveals the lights:

[v4.0.2 from v3.5.2 - Light2D Mix Blending Mode asymmetry.zip](https://github.com/godotengine/godot/files/11655458/v4.0.2.from.v3.5.2.-.Light2D.Mix.Blending.Mode.asymmetry.zip)

You can play with moving and stretching the rectangle Ground sprite. it basically stops displaying new lights after the 15th or 16th light overlapping the rectangle.
